### ISSUE TITLE
HLS: Seeking to time code?

### ISSUE THREAD

**Issue by frontyard** [google/ExoPlayer#3776]
Is there a way to seek to time code in a hls stream? In other words instead of offset from 0 to pass in actual time in millis of the recording? Does exoplayer have access / keep track of the actual time in the stream?

This would be very useful for more accurate live stream time shifting.

Here is a snippet of my HLS m3u8 file. Each fragment's file name is absolute time in millis.
 
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-MEDIA-SEQUENCE:198334
#EXT-X-TARGETDURATION:40
#EXTINF:27.680,
1517250484328.ts
#EXTINF:32.360,
1517250512278.ts
#EXTINF:35.000,
1517250544603.ts
#EXTINF:24.960,
1517250579725.ts
#EXTINF:34.120,
1517250604559.ts

**Comment by AquilesCanta** [comment]
I am not sure I understand the question, but the timeline keeps track of the window position (look at Window.positionInFirstPeriodUs, which you can obtain from onTimelineChanged). This value will be zero the first time the playlist is downloaded, and each past segment's duration will be added to it. So, to seek to an "absolute time in the stream", you'd have to seek to the position (absolute_timestamp - window_position_in_period) in the window. 

> In other words instead of offset from 0 to pass in actual time in millis of the recording

With the information in the playlist it is impossible to know when the recording started. Assuming the media sequence started at 0, the stream has been on for less than 91 days (using the target duration as upper limit). But this is just an upper limit, so it is impossible to obtain a timestamp from that.

> Each fragment's file name is absolute time in millis

The millis timestamp in years is a bit over 48 years, so I'd say the value is a timestamp relative to the epoch. However, the correct way to do this would be to add a [EXT-X-PROGRAM-DATE-TIME tag](https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-4.3.2.6). Seeking by the file timestamp is not trivial or elegant to do but there are a few ways, like passing your own playlist parser and manually converting the ts file's timestamp into a EXT-X-PROGRAM-DATE-TIME and then forwarding that to the existing parser.

To sum up, currently you can easily seek to a position from the point at which the media source started tracking the stream (starts at 0). Seeking using the file names' timestamp is a bit trickier, and preferrably fixed server-side.

**Comment by frontyard** [comment]
How does exoplayer use EXT-X-PROGRAM-DATE-TIME, is it reported somehow?

The goal is quite simple:

While zero-based position makes sense for VOD or static video files, where we can tell that playback is at 1h 14min from the beginning, for live streams (like TV, webcams etc)  it's much more useful to deal with time in absolute terms, in other words to report the actual time being played, like 2017-06-09T05:00:18.360Z instead of milli/micro seconds since the beginning of the stream. 

Beginning of live stream is not 0 but some actual time. Our current approach to do some simple math to figure out the actual time is not precise and correct. It relies on knowing how big the stream recording is which can change over time.

**Comment by frontyard** [comment]
Related question:

https://github.com/google/ExoPlayer/issues/2930

**Comment by AquilesCanta** [comment]
> How does exoplayer use EXT-X-PROGRAM-DATE-TIME, is it reported somehow?

Yes, it is exposed as `Window#positionInFirstPeriodUs` through `onTimelineChanged`.

> it's much more useful to deal with time in absolute terms, in other words to report the actual time being played, like 2017-06-09T05:00:18.360Z instead of milli/micro seconds since the beginning of the stream.

That is exactly the purpose of EXT-X-PROGRAM-DATE-TIME.
